sharktank/sharktank/utils/llm_utils.py

```diff
@@ -266,6 +268,172 @@ def allocate(self, *shape, dtype, device_index: int):
         return torch.zeros(*shape, dtype=dtype, device=self._device)
 
 
+class LlmTaskType(Enum):
+    PREFILL = auto()
+    DECODE = auto()
+
+
+@dataclasses.dataclass
+class LlmTaskInput:
+    tokens: List[int]
+    seq_len: int
+    pages: List[int]
+
+    start_position: Optional[int] = None
+
+
+class LlmTask(ABC):
+    def __init__(
+        self,
+        task_type: LlmTaskType,
+        instance: IreeInstance | TorchInstance,
+        llm_task_inputs: List[LlmTaskInput],
+        batch_size: int,
+        block_stride: int,
+    ):
+        self._task_type = task_type
+        self._instance = instance
+        self._task_inputs: List[LlmTaskInput] = llm_task_inputs
+        self._batch_size = batch_size
+        self._block_stride = block_stride
+
+    @abstractmethod
+    def _prepare_args(
+        self, task_inputs: List[LlmTaskInput], cache
+    ) -> List[numpy.ndarray | iree.runtime.DeviceArray | torch.Tensor]:
+        pass
+
+    @abstractmethod
+    def _process_results(
+        self, results
+    ) -> Tuple[numpy.ndarray, Optional[numpy.ndarray]]:
+        pass
+
+    def _fill_page_table(
```

Comment:
Can we move this to util function instead? I am being a hypocrite because I added it but seeing it as part of tasks is a bit weird considering its a static function anyway.